File: app/src/main/java/com/dsy/mvp/base/BasePresenter.java
package com.dsy.mvp.base;


import androidx.annotation.NonNull;

import com.dsy.mvp.base.impl.IPresenter;
import com.dsy.mvp.base.impl.IView;
import com.lzy.okgo.OkGo;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BasePresenter<T extends IView> implements IPresenter {
    protected T mView;
    private CompositeDisposable mCompositeDisposable;

    @Override
    public void attachView(@NonNull IView iView) {
        mView = (T) iView;
    }

    @Override
    public void detachView() {
        OkGo.getInstance().cancelTag(this);
        onUnSubscribe();
    }

    //RxJava取消注册
    public void onUnSubscribe() {
        if (mCompositeDisposable != null) {
            mCompositeDisposable.dispose();
        }
    }

    public void addSubscription(Disposable disposable) {
        if (mCompositeDisposable == null) {
            mCompositeDisposable = new CompositeDisposable();
        }
        mCompositeDisposable.add(disposable);
    }

}

File: app/src/main/java/com/dsy/mvp/base/BasePresenter.kt
package com.dsy.mvp.base

import androidx.lifecycle.*
import com.dsy.mvp.base.impl.IPresenter
import com.dsy.mvp.base.impl.IView
import com.dsy.mvp.utils.MLog
import com.lzy.okgo.OkGo
import com.rxlife.coroutine.RxLifeScope
import java.lang.ref.SoftReference


@Suppress("UNCHECKED_CAST")
abstract class BasePresenter<T : IView> : IPresenter, LifecycleObserver {
    protected val TAG = this.javaClass.simpleName
    protected val mView: T?
        get() {
            return mReferenceView?.get() as T
        }
    //弱引用view 防止内存泄露
    private var mReferenceView: SoftReference<IView>? = null

    //协程生命周期自动管理
    protected val rxLifeScope: RxLifeScope
        get() {
            return (mView as LifecycleOwner).rxLifeScope
        }

    override fun attachView(iView: IView) {
        mReferenceView = SoftReference(iView)
        if (mView is LifecycleOwner) {
            (mView as LifecycleOwner).lifecycle.addObserver(this)
        }
    }

    override fun detachView() {
        OkGo.getInstance().cancelTag(this)
        mReferenceView?.clear()
        mReferenceView = null
    }
/*

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    open fun onStart(owner: LifecycleOwner) {
        MLog.d("presenter", "$TAG.onStart()")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    open fun onResume(owner: LifecycleOwner) {
        MLog.d("presenter", "$TAG.onResume()")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    open fun onStop(owner: LifecycleOwner) {
        MLog.d("presenter", "$TAG.onStop()")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    open fun onPause(owner: LifecycleOwner) {
        MLog.d("presenter", "$TAG.onPause()")
    }
*/

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    open fun onDestroy(owner: LifecycleOwner) {
        owner.lifecycle.removeObserver(this)
//        MLog.d("presenter", "$TAG.onDestroy()")
    }
}